src/community/label_propagation.c

```diff
@@ -26,15 +26,410 @@
+                        if (k != j)
+                            running = true;
+                        /* Actually change label */
+                        VECTOR(*membership)[v1] = k;
+                    }
+                } else {
+                    /* We are not using retention, so check if we should do a control iteration
+                    or an update iteration. */
+                    if (control_iteration) {
+                        /* Check if the _current_ label of the node is also dominant */
+                        k = VECTOR(*membership)[v1];
+                        if (k < 0 ||                              /* No label assigned yet or */
+                            VECTOR(label_counters)[k] < max_count /* Label is not maximum */
+                            ) {
+                            /* Nope, we need at least one more iteration */
+                            running = true;
+                        }
+                    }
+                    else {
+                        /* Select randomly from the dominant labels */
+                        k = RNG_INTEGER(0, igraph_vector_int_size(&dominant_labels) - 1);
+                        VECTOR(*membership)[v1] = VECTOR(dominant_labels)[k];
+                    }
+                }
+            }
+
+            /* Clear the nonzero elements in label_counters */
+            num_neis = igraph_vector_int_size(&nonzero_labels);
+            for (j = 0; j < num_neis; j++) {
+                VECTOR(label_counters)[VECTOR(nonzero_labels)[j]] = 0;
+            }
+        }
+
+        /* Alternating between control iterations and label updating iterations */
+        if (!retention)
+                control_iteration = !control_iteration;
+    }
+
+    RNG_END();
+
+    if (weights) {
+        igraph_inclist_destroy(&il);
+    } else {
+        igraph_adjlist_destroy(&al);
+    }
+    IGRAPH_FINALLY_CLEAN(1);
+
+    igraph_vector_int_destroy(&node_order);
+    igraph_vector_int_destroy(&nonzero_labels);
+    igraph_vector_int_destroy(&dominant_labels);
+    igraph_vector_destroy(&label_counters);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_i_community_fast_label_propagation(const igraph_t *graph,
+    igraph_vector_int_t *membership,
+    igraph_neimode_t mode,
+    const igraph_vector_t *weights,
+    igraph_vector_bool_t *fixed) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph), no_of_not_fixed_nodes = 0;
+    igraph_integer_t i, j, k;
+    igraph_inclist_t il;
+    igraph_adjlist_t al;
+
+    igraph_vector_t label_counters;
+    igraph_vector_int_t dominant_labels, nonzero_labels, node_order;
+    igraph_dqueue_t queue;
+    igraph_vector_bool_t in_queue;
+    igraph_neimode_t reverse_mode;
+
+    reverse_mode = IGRAPH_REVERSE_MODE(mode);
+
+    if (weights) {
+        IGRAPH_CHECK(igraph_inclist_init(graph, &il, reverse_mode, IGRAPH_LOOPS_ONCE));
```

Comment:
I know that the `LOOPS_ONCE` choice was not made in this PR, but do we have a reasoning for using it?

Almost everywhere else, we use `LOOPS_TWICE`, unless there is a clear reason not to. In order for igraph to behave predictably, it makes sense to me to do the same here. Since your understanding of label propagation is fresh at the moment, @vtraag, do you see a reason to keep this `LOOPS_ONCE` on `develop`?

Reply:
I think there's no strong argument for either. Self-loops in this case would matter for how strong the effect is on keeping the same label. For example, suppose we have a node with four neighbours and a self-loop, and two neighbours have label A and two have label B, while the node's own label is label C. Then with `LOOPS_ONCE` label C would not be dominant (i.e. a maximal label), because label C is counted only once with a self-loop, while with `LOOPS_TWICE` is would be. I'm not sure what the original paper suggested (if anything), but to me `LOOPS_ONCE` actually seems more reasonable than `LOOPS_TWICE` here.